luci-mod-status: add simple filtering for connections

McGiverGim opened the pull request:
This commit adds a simple filtering input text into the connections status page. You can type some filtering like:
- '100.15' -> Will show only connections that include this '100.15' onto the IP.
- 'UDP 100.15' -> Will show only UDP connections that include this '100.15' onto the IP.

You can use a part of the IP or part of the name if the DNS resolution button is enabled.

It's a very simple filter that can grow in the future if it deserves some interest, maybe adding some regular expression support.

![image](https://github.com/user-attachments/assets/631d11d4-8e86-4690-a4d5-af5598bbc409)

systemcrash commented on modules/luci-mod-status/htdocs/luci-static/resources/view/status/connections.js:
```diff
@@ -359,23 +367,48 @@ return view.extend({
+						E('input', {
+							id: 'filter-connections',
+							type: 'text',
+							placeholder: _('Type some filtering for connections...'),
```
Perhaps revise to something more 'useful' - like an example search?

jow- replied:
Also please use `…` (&hellip;) instead of three ordinary dots.

McGiverGim replied:
I was thinking the same, but it seemed too long to me to put examples, specially for small screens. But I can change it, maybe something short like: 
"Filter like '0.15' or 'tcp 0.15'”
I'm not English native, so any help will be appreciated...

systemcrash replied:
How about something simple like `UDP 192.168.1.1`? Then i18n is not necessary.

McGiverGim replied:
The field does not have any "label" indicating what is this for, so for this reason I added the "filter" word to the placeholder. But if you think it's enough adding the sample without text, I can change it, no problem.

McGiverGim replied:
Do I change this finally? If yes, what about different filters to show different samples with a separator?
`192.168.1.15 / UDP 1.15 / :443`

Maybe it's confusing...

systemcrash replied:
That’s great

McGiverGim replied:
Ok, changed as discused.

McGiverGim commented:
I replaced the text with the samples (but I'm ready to change it if someone prefers a better wording) and added comparison with ignore case to protocol and network.

systemcrash commented on modules/luci-mod-status/htdocs/luci-static/resources/view/status/connections.js:
```diff
@@ -130,12 +131,27 @@ return view.extend({
 			var src = dns_cache[c.src] || (c.layer3 == 'ipv6' ? '[' + c.src + ']' : c.src);
 			var dst = dns_cache[c.dst] || (c.layer3 == 'ipv6' ? '[' + c.dst + ']' : c.dst);
 
+			let network = c.layer3.toUpperCase();
+			let protocol = c.layer4.toUpperCase();
+			let source ='%h'.format(c.hasOwnProperty('sport') ? (src + ':' + c.sport) : src);
+			let destination = '%h'.format(c.hasOwnProperty('dport') ? (dst + ':' + c.dport) : dst);
+			let transfer = [ c.bytes, '%1024.2mB (%d %s)'.format(c.bytes, c.packets, _('Pkts.')) ];
```
consider:
```js
const network = c.layer3?.toUpperCase() || '';
const protocol = c.layer4?.toUpperCase() || '';
const source = c.hasOwnProperty('sport') ? `${src}:${c.sport}` : src;
const destination = c.hasOwnProperty('dport') ? `${dst}:${c.dport}` : dst;
const transfer = [ c.bytes, `${(c.bytes / 1024 / 1024).toFixed(2)} MB (${c.packets} ${_('Pkts.')})` ];
```

McGiverGim replied:
True, I will change it.

systemcrash commented on modules/luci-mod-status/htdocs/luci-static/resources/view/status/connections.js:
```diff
@@ -130,12 +131,27 @@ return view.extend({
+
+			if  (filterText) {
+				let filterTextExpressions = filterText.split(' ');
+				if (filterTextExpressions.some((element) => element.toUpperCase !== network && element.toUpperCase() !== protocol 
```
Should that be a function?

McGiverGim replied:
In reallity it must return a 'truthy' value https://developer.mozilla.org/en-US/docs/Glossary/Truthy

You can use the complete function or like here, an arrow function: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some#examples

systemcrash replied:
You have `element.toUpperCase !== network` (not a function call)
but you have `element.toUpperCase() !== protocol` (function call)

McGiverGim replied:
True, this is a bug :( I will change it. Good catch! I didn't understand your first explanation, sorry...

McGiverGim replied:
Fixed!

systemcrash commented on modules/luci-mod-status/htdocs/luci-static/resources/view/status/connections.js:
```diff
@@ -130,12 +131,27 @@ return view.extend({
+			if  (filterText) {
+				let filterTextExpressions = filterText.split(' ');
+				if (filterTextExpressions.some((element) => element.toUpperCase !== network && element.toUpperCase() !== protocol 
+						&& !(c.src.includes(element) || source.includes(element))
+						&& !(c.dst.includes(element) || destination.includes(element)))) {
```
Can we skip the c.* checks? They're implied in `source` and `destination`, yes?

McGiverGim replied:
Not exactly... 
- the `c.src` and `c.dst` contains the IP always
- the `source` and `destination` contains the IP or the HOSTNAME (if the DNS resolution button is enabled) and the PORT

So this is a simple way to filter by IP or hostname. If the DNS resolution is disabled it will filter by IP twice, but if I try to represent this in the condition in the filter to use it only when the DNS resolution is disabled it will be a more complicated function, so I decided to let it simple.

systemcrash commented:
It’s low effort, and high value when dealing with lots of endpoints, so it’s a welcome addition I think. @McGiverGim also thinks so.

McGiverGim commented:
I think it's useful when you want to see the connections from a host, for example. If this is not needed, then the full list of connections it's not needed neither, because now it's a big list but you can't find anything because it's big and changes continuosly.

systemcrash commented:
Took it for a spin. Works fine.

systemcrash commented:
Thanks @McGiverGim
